### PR TITLE
parser: parse gbk encoding SQL token by token instead of the whole SQL

### PR DESCRIPTION
**Pull request by xiongjiwei**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/30765
close https://github.com/pingcap/tidb/issues/29382

this pr is based on https://github.com/pingcap/tidb/pull/29573. please only review the second commit.
It will make the parser compatible with MySQL in parsing gbk encoding characters.

```go
// benchcode
func BenchmarkParser(b *testing.B) {
	p := parser.New()
	sql := "select '" + strings.Repeat("这是一个测试", 10000) + "' as " + strings.Repeat("aaaaa", 10000)
	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		_, _, err := p.ParseSQL(sql)
		require.NoError(b, err)
	}
}
```

`client=utf8 & connection=utf8`
```
before                       after
876	   1410416 ns/op     775	   1497224 ns/op
829	   1427232 ns/op     831	   1486620 ns/op
686	   1570810 ns/op     660	   1678825 ns/op
747	   1512891 ns/op     622	   1718708 ns/op
781	   1512988 ns/op     706	   1662557 ns/op
```
image

`client=gbk & connection=utf8`
`client=gbk & connection=gbk`
```
240	   5076879 ns/op
237	   5216370 ns/op
232	   5058759 ns/op
234	   5134991 ns/op
234	   5142602 ns/op
```
image

we need optimize `convert2Connection` function
### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

### Release note

```release-note
None
```

### CODE REVIEW

File: parser/charset/encoding_base.go [hunk 1]
```diff
@@ -34,6 +34,10 @@ type encodingBase struct {
 	self Encoding
 }
 
+func (b encodingBase) MbLen(_ string) int {
+	return 0
+}
+
```
Comment (tangenta) [inline parser/charset/encoding_base.go:40]:
```suggestion
```
It is better not to implement in `encodingBase` because the parent may forget this.

Comment (xiongjiwei) [reply parser/charset/encoding_base.go:40]: I think this is what I need, it is the default implement

File: parser/charset/encoding_gbk.go [hunk 2]
```diff
@@ -55,6 +55,20 @@ func (e *encodingGBK) Peek(src []byte) []byte {
 	return src
 }
 
+func (e *encodingGBK) MbLen(bs string) int {
+	if len(bs) < 2 {
+		return 0
+	}
+
+	if 0x81 <= bs[0] && bs[0] <= 0xf4 {
```
Comment (tangenta) [inline parser/charset/encoding_gbk.go:63]: Could you add a comment/link about 2 bytes GBK characters?

File: parser/lexer.go [hunk 3]
```diff
@@ -145,19 +147,20 @@ func (s *Scanner) AppendWarn(err error) {
 	s.warns = append(s.warns, err)
 }
 
-func (s *Scanner) tryDecodeToUTF8String(sql string) string {
-	if mysql.IsUTF8Charset(s.encoding.Name()) {
-		// Skip utf8 encoding because `ToUTF8` validates the whole SQL.
-		// This can cause failure when the SQL contains BLOB values.
-		// TODO: Convert charset on every token and use 'binary' encoding to decode token.
-		return sql
+func (s *Scanner) convert2Connection(tok int, sql string) (int, string) {
```
Comment (tangenta) [inline parser/lexer.go:150]:
```suggestion
func (s *Scanner) convert2Connection(tok int, lit string) (int, string) {
```

File: parser/lexer.go [hunk 4]
```diff
@@ -926,39 +958,25 @@ func (s *Scanner) lastErrorAsWarn() {
 type reader struct {
 	s string
 	p Pos
-	w int
+	l int
```
Comment (tangenta) [inline parser/lexer.go:961]: Why do we need this instead of using `len(s)` directly?

Comment (xiongjiwei) [reply parser/lexer.go:961]: `len(s)` is much slower than a const

File: parser/yy_parser.go [hunk 5]
```diff
@@ -293,7 +292,7 @@ func toFloat(l yyLexer, lval *yySymType, str string) int {
 			return int(unicode.ReplacementChar)
```
Comment (tangenta) [inline parser/yy_parser.go:292]: Please change `unicode.ReplacementChar` to `invalid`.

File: parser/lexer.go [hunk 6]
```diff
@@ -993,25 +1010,32 @@ func (r *reader) pos() Pos {
...
-		if ch == unicode.ReplacementChar && r.eof() {
+		if r.eof() {
 			return 0
 		}
 		r.inc()
 	}
 }
+
+// SkipRune skip mb character, return true indicate something has been skipped.
+func (r *reader) SkipRune() bool {
```
Comment (tangenta) [inline parser/lexer.go:1034]: How about using `Encoding` as the parameter? The function pointer `f` in `reader` is not needed anymore.

BTW, this function can be unexported.

Comment (xiongjiwei) [reply parser/lexer.go:1034]: great

File: parser/parser_test.go [hunk 7]
```diff
@@ -6413,18 +6413,29 @@ func TestGBKEncoding(t *testing.T) {
...
+		err bool
+	}{
+		{"select '\xc6\x5c' from `\xab\x60`;", false},
+		{`prepare p1 from "insert into t values ('中文');";`, false},
+		{"select '啊';", false},
+		{"create table t1(s set('a一','b二','c三'));", false},
+		{"insert into t3 values('一a');", false},
+		{"select '\xa5\x5c'", false},
+		{"select '''\xa5\x5c'", false},
+		{"select ```\xa5\x5c`", false},
```
Comment (tangenta) [inline parser/parser_test.go:6430]: Could you please add some `err == true` cases?

Comment (xiongjiwei) [reply parser/parser_test.go:6430]: updated

File: parser/charset/encoding_gbk.go (hunk 2 above)
Comment (xiongjiwei) [reply parser/charset/encoding_gbk.go:63]: the 2 bytes character can refer to https://github.com/mysql/mysql-server/blob/3290a66c89eb1625a7058e0ef732432b6952b435/strings/ctype-gbk.cc#L3605

File: parser/lexer.go [hunk 8]
```diff
@@ -88,7 +89,9 @@ func (s *Scanner) Errors() (warns []error, errs []error) {
 
 // reset resets the sql string to be scanned.
 func (s *Scanner) reset(sql string) {
-	s.r = reader{s: sql, p: Pos{Line: 1}}
+	s.client = charset.EncodingUTF8Impl
```
Comment (Defined2014) [inline parser/lexer.go:92]: How about `s.client = charset.FindEncoding(mysql.DefaultCharset)`?

Comment (tangenta) [reply parser/lexer.go:92]: I think we can use `charset.EncodingBinaryImpl` instead and remove line 151.

File: parser/lexer.go [hunk 9]
```diff
@@ -144,19 +147,20 @@ func (s *Scanner) AppendWarn(err error) {
...
+	utf8Lit, err := s.client.Transform(nil, charset.Slice(sql), charset.OpDecodeReplace)
 	if err != nil {
 		s.AppendError(err)
+		if s.sqlMode.HasStrictMode() && s.client.Tp() == s.connection.Tp() {
+			return invalid, sql
+		}
 		s.lastErrorAsWarn()
 	}
-	return string(utf8Lit)
+	utf8Lit, _ = s.connection.Transform(nil, utf8Lit, charset.OpReplace)
```
Comment (Defined2014) [inline parser/lexer.go:162]: Why we need this? Could you add a test case to figure it?
Seems if `s.client.Tp() == s.connection.Tp()` we could skip it?

Comment (xiongjiwei) [reply parser/lexer.go:162]: MySQL will convert tokens from `client` to `connection`. (see [doc](https://dev.mysql.com/doc/refman/8.0/en/charset-connection.html#:~:text=The%20server%20converts%20statements%20sent%20by%20the%20client%20from%20character_set_client%20to%20character_set_connection)). if `connection` is not `utf8mb4`, we should replace the invalid character with `?` here. 

if `client` and `connection` is the same, we cannot be compatible with MySQL, see the solution https://github.com/pingcap/tidb/issues/29382

**Review by Defined2014 — APPROVED** [review]
Please update gbk performance @xiongjiwei , others LGTM

**Review by tangenta — APPROVED** [review]

**Pull request merged**